streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -432,24 +490,54 @@ private void cleanupTask(final Task task) {
 
     void shutdown(final boolean clean) {
         final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
-        final Iterator<Task> iterator = tasks.values().iterator();
-        while (iterator.hasNext()) {
-            final Task task = iterator.next();
+
+        final Map<Task, Map<TopicPartition, Long>> checkpointPerTask = new HashMap<>();
+        final Map<TaskId, Map<TopicPartition, OffsetAndMetadata>> consumedOffsetsAndMetadataPerTask = new HashMap<>();
+        final Set<Task> dirtyTasks = new HashSet<>();
+
+        for (final Task task : tasks.values()) {
             cleanupTask(task);
 
             if (clean) {
                 try {
-                    task.closeClean();
+                    checkpointPerTask.put(task, task.prepareCloseClean());
+                    if (task.state() != CREATED) {
+                        consumedOffsetsAndMetadataPerTask.put(task.id(), task.committableOffsetsAndMetadata());
+                    }
                 } catch (final TaskMigratedException e) {
                     // just ignore the exception as it doesn't matter during shutdown
-                    task.closeDirty();
+                    task.prepareCloseDirty();
+                    dirtyTasks.add(task);
                 } catch (final RuntimeException e) {
                     firstException.compareAndSet(null, e);
-                    task.closeDirty();
+                    task.prepareCloseDirty();
+                    dirtyTasks.add(task);
                 }
             } else {
+                task.prepareCloseDirty();
+                dirtyTasks.add(task);
+            }
+        }
+
+        if (clean && !consumedOffsetsAndMetadataPerTask.isEmpty()) {
+            commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
+        }
+
+        for (final Map.Entry<Task, Map<TopicPartition, Long>> taskAndCheckpoint : checkpointPerTask.entrySet()) {
+            final Task task = taskAndCheckpoint.getKey();
+            try {
+                task.closeClean(checkpointPerTask.get(task));
+            } catch (final RuntimeException e) {
+                firstException.compareAndSet(null, e);
                 task.closeDirty();
             }
+        }
+
+        for (final Task task : dirtyTasks) {
```

Comment:
Do we need to keep a task once it is failed to clean close? Why couldn't we just close it dirty immediately after we see the exception?

Reply:
Good idea!